apps/labs/posts/numpy-string-ufuncs.md

```diff
@@ -0,0 +1,546 @@
+with just one input and one output, the loop was relatively easy to write. It looked
+something like this[^1]:
+
+```cpp
+void
+string_isalpha_loop(char **args, npy_intp const *dimensions, npy_intp const *steps, void *data)
+{
+    // Get a pointer to the input and output data buffers
+    char *in = args[0];
+    char *out = args[1];
+
+    // Get number of items to loop over
+    npy_intp N = dimensions[0];
+
+    while (N--) {
+        npy_bool res = string_isalpha(in);
+
+        // Write the result to the output buffer
+        *(npy_bool *)out = res;
+
+        // Jump to the next element in the array
+        in += steps[0];
+        out += steps[1];
+    }
+}
+```
+
+Within the loop function, we do the following:
+
+1. Get a pointer to the input and the output data buffers using the `args` argument.
+2. Loop over all of the items in the arrays by using the `dimensions` argument,
+   and, for each item, do the following:
+   - Use the input buffer to call `string_isalpha` and get a result back. Note that
+     we do not include the implementation of `string_isalpha` here, because the
+     implementation itself is irrelevant to the ufunc mechanism.
+   - Write the result to the output buffer.
+   - Jump to the next element in the array by adjusting the input and output
+     pointers using the `steps` argument.
+
+The results from writing a ufunc for `isalpha` showed that writing ufuncs really
+was the way to go (more on this later), so from this point on, the real work of
+writing ufuncs for *all* the string operations began.
+
+## A brief digression: Fixed-length string dtypes
+
+In order to understand the rest of the article, it's important to have an idea of
+how strings are represented in memory in NumPy. NumPy, as of NumPy 1.26, has two
+string dtypes, one for Unicode (UTF-32) and one for bytes (ASCII). Both of these
+dtypes are fixed-length dtypes, which means that the maximum length of each array
+item is specified by the dtype. Let's take the following example:
+
+```python3
+>>> import numpy as np
+>>> s = np.array(["abcd"])
+>>> s
+array(['abcd'], dtype='<U4')
+```
+
+The `dtype='<U4'` tells us that the strings in this array have a maximum length
+of four characters. Were we to assign an item of the array to a string that's
+longer than that, the excess characters would be stripped:
+
+```python3
+>>> s[0] = 'efghi'
+>>> s
+array(['efgh'], dtype='<U4')
+```
+
+That doesn't mean, however, that all of the array items have to be exactly four
+characters long. We can have array items that are smaller than that. The underlying
+array buffer still allocates the number of bytes corresponding to four characters
+(4 bytes for ASCII, 16 bytes for UTF-32), but the rest of the bytes are zero. That
+has the side-effect that both of these dtypes strip all trailing zero bytes.
+
+```python3
+>>> s = np.array(["abcd", "ef"])
+>>> s
+array(['abcd', 'ef'], dtype='<U4')
+>>> s[0] = "h\0\0"
+>>> s
+array(['h', 'ef'], dtype='<U4')
```

Comment:
you can illustrate how the actual array is laid out in memory printing the result of `s.tobytes()`, I sometimes find that useful to understand this.